activerecord/lib/active_record/relation/batches.rb

```diff
@@ -227,35 +230,36 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
         batch_limit = remaining if remaining < batch_limit
       end
 
-      relation = relation.reorder(batch_order(order)).limit(batch_limit)
-      relation = apply_limits(relation, start, finish, order)
+
+      relation = relation.reorder(batch_order(order, order_column)).limit(batch_limit)
+      relation = apply_limits(relation, start, finish, order, order_column)
       relation.skip_query_cache! # Retaining the results in the query cache would undermine the point of batching
       batch_relation = relation
       empty_scope = to_sql == klass.unscoped.all.to_sql
 
       loop do
         if load
           records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
+          ids = records.map { |r| r.send(order_column) }
+          yielded_relation = where(order_column => ids)
           yielded_relation.load_records(records)
         elsif (empty_scope && use_ranges != false) || use_ranges
-          ids = batch_relation.pluck(primary_key)
+          ids = batch_relation.pluck(order_column)
           finish = ids.last
           if finish
-            yielded_relation = apply_finish_limit(batch_relation, finish, order)
+            yielded_relation = apply_finish_limit(batch_relation, finish, order, order_column)
             yielded_relation = yielded_relation.except(:limit, :order)
             yielded_relation.skip_query_cache!(false)
           end
         else
-          ids = batch_relation.pluck(primary_key)
-          yielded_relation = where(primary_key => ids)
+          ids = batch_relation.pluck(order_column)
```

Comment:
Looking at this line, I think that `order_column` is slightly wrong name for this as an **identifier** is what we actually want to be able to configure which by implication should dictate the `order by` clause. 

Also naming this configuration closer to something like an `identifier` should reveal the need for the `identifier` to be `unique`, and I would add an explicit check for that that raises an exception if the configured column is not unique.  Otherwise treating a non-unique column as the `cursor` may lead to records being skipped when the next batch loads records for the next unique value of the `order_column`.

It also made me think whether batching methods should let the model itself to define an alternative identifier to be used in batching finders. Similar to the existing `implicit_order_column` config. Don't think `implicit_order_column` can be used though as it doesn't seem to imply uniqueness of the column

Reply:
Hi @nvasilevski thanks for your comment.

I understand that the column has to be unique and a check could be ok to avoid weird behaviour, is it possible to check that?

In addition, not sure if identifier it is a good name. Maybe I’m not aware of some active record internals but in this context I think order_column name fits well for the purpose.

About the config in the model, could be possible but isn’t it better to specify that sorting column in these methods (explicitly) instead of polluting the model with a “global configuration”?

Reply:
> I think order_column name fits well for the purpose.

I'm afraid it doesn't really fit as `ids = batch_relation.pluck(order_column)` becomes a very confusing line and `ids` becomes an array of values the relation was ordered which is fair from being an array of ids. 
I'm highly confident that the `order_column` is too narrow because as shown in the description the new argument changes not just the `ORDER BY` clause but basically whole query, including the "cursor" condition: `WHERE "posts"."version" > ?`  
I agree that `identifier` is not a good name, but I'm trying to find something that will imply that we are configuring the batching behavior overall and not just the ordering part. Perhaps `cursor_column`? Or `cursor_identifier` 
Also, if you combine the usage of the proposed `order_column` with `use_ranges: true` argument, there won't be `ORDER BY` even though the `order_column` will be used as a column to build ranges

Reply:
>I understand that the column has to be unique and a check could be ok to avoid weird behaviour, is it possible to check that?

I would wait for another opinion on that, perhaps such check would be too expensive to have and it would be better just to have it written as a general warning in the docs. But this check may look like: the column is a primary key or has a unique index. The `has a unique index` piece may be expensive or at least look excessive

Reply:
@nvasilevski I've finally replaced with `cursor_column` name :)

Reply:
Anytime @nvasilevski ;)